draft-ietf-quic-load-balancers.md

```diff
@@ -232,97 +258,164 @@ Config Rotation: Indicates the configuration used to interpret the CID.
-QUIC version 1, this requires using a routable CID in the Source CID field of
-server-generated long headers.
+For packets not known to belong to a QUIC version the load balancer can parse,
+load balancers MUST use the baseline fallback algorithm if the DCID is
+unroutable.
 
-## Fallback Algorithms {#fallback-algorithm}
+For known QUIC versions, the fallback algorithm MAY parse packets and use that
+information to make a routing decision.
+
+If so, it MUST have the ability to buffer packets with unroutable DCIDs to await
+further packets that allow it to make a routing decision, as the fields of
+interest can be an arbitary number of packets into the connection.
+
+4-tuple routing is not sufficient for this use case, because a client can use
+the same 4-tuple for two connections that should be routed differently (e.g.
+because they target different SNIs), as long as the packet contains a source
+connection ID of nonzero length.
 
-There are conditions described below where a load balancer routes a packet using
-a "fallback algorithm." It can choose any algorithm, without coordination with
-the servers, but the algorithm SHOULD be deterministic over short time scales so
-that related packets go to the same server.
+Therefore, the load balancer SHOULD maintain two tables that map different
+values to a routing decision:
 
-A fallback algorithm MAY parse a packet from a QUIC version it understands and
-use that information to make a routing decision.
+- a table indexed by a concatenation of the 4-tuple and source CID, which might
+be zero-length, to route subsequent packets that do not contain the server-
+generated connection ID;
 
-For versions it does not understand, a fallback algorithm MUST NOT make the
-routing behavior dependent on any part of the long header that is not invariant
-across versions (see {{!RFC8999}}).
+- a table indexed by destination CID, if and only if it is possible for the
+server to generate unroutable CIDs. This table can be shared with the one in use
+for the baseline fallback algorithm.
 
-For example, one fallback algorithm might convert a unroutable DCID to an
-integer and divided by the number of servers, with the modulus used to forward
-the packet. The number of servers is usually consistent on the time scale of a
-QUIC connection handshake. Another might simply hash the address/port 4-tuple.
-See also {{version-invariance}}.
+If either table overflows, or if the load balancer loses state, it is likely the
+load balancer will misroute packets.
+
+Load balancers SHOULD maintain per-flow timers to periodically purge state in
+the tables described above.
+
+## Dropping packets
+
+Some unroutable connection IDs are evidently client-generated if all servers
+behind the load balancer are compliant with this specification. In particular,
+an unroutable CID is evidently client-generated if
+- it does not use the 0b111 config ID; OR
+- it uses the 0b111 config ID, it is in a long header packet, and its self-
```

Comment:
Another case: if the server is encoding length, and the encoded length is known to not be a valid CID length the server would generate